cli/tools/check.rs

```diff
@@ -34,12 +46,156 @@ use crate::tsc;
+    ConfigFlag::Discover => true,
+    ConfigFlag::Path(_) => false,
+    ConfigFlag::Disabled => false,
+  };
+  if is_discovered_config {
+    let factory = CliFactory::from_flags(flags.clone());
+    let cli_options = factory.cli_options()?;
+    let (remote_files, files) = check_flags
+      .files
+      .iter()
+      .cloned()
+      .partition::<Vec<_>, _>(|f| {
+        f.starts_with("http://")
+          || f.starts_with("https://")
+          || f.starts_with("npm:")
+          || f.starts_with("jsr:")
+      });
+    let mut by_workspace_directory = cli_options
+      .resolve_file_flags_for_members(&FileFlags {
+        ignore: Default::default(),
+        include: files,
+      })?
+      .into_iter()
+      .map(|(d, p)| (d.dir_url().clone(), (Arc::new(d), p)))
+      .collect::<BTreeMap<_, _>>();
+    if !remote_files.is_empty() {
+      by_workspace_directory
+        .entry(cli_options.start_dir.dir_url().clone())
+        .or_insert((
+          cli_options.start_dir.clone(),
+          FilePatterns {
+            base: cli_options.initial_cwd().to_path_buf(),
+            include: Some(Default::default()),
+            exclude: Default::default(),
+          },
+        ))
+        .1
+        .include
+        .get_or_insert_with(Default::default)
+        .append(
+          remote_files
+            .iter()
+            .flat_map(|s| ModuleSpecifier::parse(s).ok())
+            .map(PathOrPattern::RemoteUrl),
+        );
+    }
+    let all_scopes = Arc::new(
+      by_workspace_directory
+        .iter()
+        .filter(|(_, (d, _))| d.has_deno_or_pkg_json())
+        .map(|(s, (_, _))| s.clone())
+        .collect::<BTreeSet<_>>(),
+    );
+    let dir_count = by_workspace_directory.len();
+    let mut diagnostics = vec![];
+    let mut all_errors = vec![];
+    let mut found_specifiers = false;
+    for (dir_url, (workspace_directory, patterns)) in by_workspace_directory {
+      let (npmrc, _) =
+        discover_npmrc_from_workspace(&workspace_directory.workspace)?;
+      let lockfile =
+        CliLockfile::discover(&flags, &workspace_directory.workspace)?;
+      let scope_options = (dir_count > 1).then(|| ScopeOptions {
+        scope: workspace_directory
+          .has_deno_or_pkg_json()
+          .then(|| dir_url.clone()),
+        all_scopes: all_scopes.clone(),
+      });
+      let cli_options = CliOptions::new(
+        flags.clone(),
+        cli_options.initial_cwd().to_path_buf(),
+        lockfile.map(Arc::new),
+        npmrc,
+        workspace_directory,
+        false,
+        scope_options.map(Arc::new),
+      )?;
+      let specifiers = collect_specifiers(
+        patterns,
+        cli_options.vendor_dir_path().map(ToOwned::to_owned),
+        |e| is_script_ext(e.path),
+      )?;
+      if specifiers.is_empty() {
+        continue;
+      } else {
+        found_specifiers = true;
+      }
+      let factory = CliFactory::from_cli_options(Arc::new(cli_options));
```

Comment:
Ideally we should only be creating a single factory. Why is all this setup code necessary here?

Reply:
The factory's services used below are based on  `WorkspaceDirectory` currently. I wanted to contain the changes somewhere for now but I can keep working on it.